### ISSUE TITLE
关于 MoE 版本的辅助损失函数.

### ISSUE THREAD

**Issue by Leiyi-Hu** [jingyaogong/minimind#110]
您好，这里的 aux_loss 看起来并没有被使用？还是通过其他的方式参与了训练呢？
![image](https://github.com/user-attachments/assets/d3e21338-a2d3-462c-9fdc-b43f1917423f)

**Comment by jingyaogong** [comment]
是的，简单起见这部分loss并没有加入训练😊

**Comment by Leiyi-Hu** [comment]
> 是的，简单起见这部分loss并没有加入训练😊

谢谢！如果需要加入训练，其实现是不是应该将每层的 loss 都存下来和最后的 ce loss 一起进行梯度计算呢？

**Comment by jingyaogong** [comment]
> > 是的，简单起见这部分loss并没有加入训练😊
> 
> 谢谢！如果需要加入训练，其实现是不是应该将每层的 loss 都存下来和最后的 ce loss 一起进行梯度计算呢？

是的，只需要把每一层的 `aux_loss` 累加，最后和 `logits分类交叉熵loss` 相加即可。

**Comment by Leiyi-Hu** [comment]
> > > 是的，简单起见这部分loss并没有加入训练😊
> > 
> > 
> > 谢谢！如果需要加入训练，其实现是不是应该将每层的 loss 都存下来和最后的 ce loss 一起进行梯度计算呢？
> 
> 是的，只需要把每一层的 `aux_loss` 累加，最后和 `logits分类交叉熵loss` 相加即可。

明白了，谢谢！另外有一个关于数据预处理问题想请教，
![image](https://github.com/user-attachments/assets/2d3421b9-c9e9-4bc7-b763-a777344a44d6)
这里的 history 为什么截断为 max_length的一半？同时比较疑惑的是 max_length应该是以 token 为单位，这里的 history 看起来是字符串？

**Comment by jingyaogong** [comment]
> > > > 是的，简单起见这部分loss并没有加入训练😊
> > > 
> > > 
> > > 谢谢！如果需要加入训练，其实现是不是应该将每层的 loss 都存下来和最后的 ce loss 一起进行梯度计算呢？
> > 
> > 
> > 是的，只需要把每一层的 `aux_loss` 累加，最后和 `logits分类交叉熵loss` 相加即可。
> 
> 明白了，谢谢！另外有一个关于数据预处理问题想请教， ![image](https://private-user-images.githubusercontent.com/66046309/401419135-2d3421b9-c9e9-4bc7-b763-a777344a44d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0MDUxMTksIm5iZiI6MTczNjQwNDgxOSwicGF0aCI6Ii82NjA0NjMwOS80MDE0MTkxMzUtMmQzNDIxYjktYzllOS00YmM3LWI3NjMtYTc3NzM0NGE0NGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDA2NDAxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxODc1NjY5NDNiMmEwMDU1YWJjMmE5MzE0YWIyNWRhODNiMjdkNWI1MWM4ZjQwYjk4Y2FhNjhiY2IwOWJiZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Pr9fvdWQgvFSXIRxh1i_Nj3K2djLqKEopZ11SAtgvxk) 这里的 history 为什么截断为 max_length的一半？同时比较疑惑的是 max_length应该是以 token 为单位，这里的 history 看起来是字符串？

也是简单起见
如果在这里用tokenzier去严格统计token数量会增加不必要的时间（当然对后面的input_id做统计也行，只不过更麻烦）
考虑到M的字符串中的token数量一定是 < M 的

**Comment by Leiyi-Hu** [comment]
> > > > > 是的，简单起见这部分loss并没有加入训练😊
> > > > 
> > > > 
> > > > 谢谢！如果需要加入训练，其实现是不是应该将每层的 loss 都存下来和最后的 ce loss 一起进行梯度计算呢？
> > > 
> > > 
> > > 是的，只需要把每一层的 `aux_loss` 累加，最后和 `logits分类交叉熵loss` 相加即可。
> > 
> > 
> > 明白了，谢谢！另外有一个关于数据预处理问题想请教， ![image](https://private-user-images.githubusercontent.com/66046309/401419135-2d3421b9-c9e9-4bc7-b763-a777344a44d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY0MDUxMTksIm5iZiI6MTczNjQwNDgxOSwicGF0aCI6Ii82NjA0NjMwOS80MDE0MTkxMzUtMmQzNDIxYjktYzllOS00YmM3LWI3NjMtYTc3NzM0NGE0NGQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA5VDA2NDAxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxODc1NjY5NDNiMmEwMDU1YWJjMmE5MzE0YWIyNWRhODNiMjdkNWI1MWM4ZjQwYjk4Y2FhNjhiY2IwOWJiZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Pr9fvdWQgvFSXIRxh1i_Nj3K2djLqKEopZ11SAtgvxk) 这里的 history 为什么截断为 max_length的一半？同时比较疑惑的是 max_length应该是以 token 为单位，这里的 history 看起来是字符串？
> 
> 也是简单起见 如果在这里用tokenzier去严格统计token数量会增加不必要的时间（当然对后面的input_id做统计也行，只不过更麻烦） 考虑到M的字符串中的token数量一定是 < M 的

感谢您的解答！🙏
